### PR TITLE
start: wait until pods ready

### PR DESCRIPTION
**Pull request by sttts**
Minimal alternative to https://github.com/kubernetes-incubator/bootkube/pull/1029.

Processes that are accessed via a service are not serving requests until they are "ready". In the case of the kube-apiserver being ready is reported through the `/healthz` endpoint.

Without this PR, bootkube switches to a self-hosted kube-apiserver as soon as the pod is marked running. It tears down the bootstrap control plane immediately, leaving the cluster without an API server, possibly for a number of seconds.

The ready condition in the Pod is meant exactly for this purpose, i.e. to report when a pod is ready to serve. We should use it.

### CODE REVIEW

File: pkg/bootkube/status.go [hunk 1]
```diff
@@ -93,20 +89,40 @@ func (s *statusController) AllRunning() (bool, error) {
...
 
-	running := true
+	runningAndReady := true
 	for p, s := range ps {
 		if changed {
-			UserOutput("\tPod Status:%24s\t%s\n", p, s)
+			var status string
+			switch {
+			case s == nil:
+				status = "DoesNotExist"
```
Comment (mfojtik) [inline pkg/bootkube/status.go]: can we convert these to consts?

Comment (mfojtik) [reply pkg/bootkube/status.go]: actually if this is only for nice output, don't bother ;-)

File: pkg/bootkube/status.go [hunk 2]
```diff
@@ -93,20 +89,40 @@ func (s *statusController) AllRunning() (bool, error) {
...
-	return running, nil
+	return runningAndReady, nil
 }
 
-func (s *statusController) PodStatus() (map[string]v1.PodPhase, error) {
-	status := make(map[string]v1.PodPhase)
+// PodStatus describes a pod's phase and readiness.
+type PodStatus struct {
+	Phase v1.PodPhase
+	Ready bool
```
Comment (mfojtik) [inline pkg/bootkube/status.go]: nit: IsReady

File: pkg/bootkube/status.go [hunk 3]
```diff
@@ -122,11 +138,18 @@ func (s *statusController) PodStatus() (map[string]v1.PodPhase, error) {
 			return nil, err
 		}
 		if !exists {
-			status[watchedPod] = doesNotExist
+			status[watchedPod] = nil
```
Comment (mfojtik) [inline pkg/bootkube/status.go]: nit: `delete(status, watchedPod)` ?

**Comment by mfojtik** [comment]
LGTM (just nits)

File: pkg/bootkube/status.go (hunk 3 above)
Comment (sttts) [reply pkg/bootkube/status.go]: then the loop above gets more complicated. Not worth the change.

File: pkg/bootkube/status.go (hunk 2 above)
Comment (sttts) [reply pkg/bootkube/status.go]: done

**Comment by dghubble** [comment]
What's an example case for needing to enforce this at bootkube start? Can you expand the description please?

**Comment by sttts** [comment]
@dghubble extended the description. tl/dr: a self-hosted control plane is not able to serve until the kube-apiserver pod is reported to be ready (controlled via a `/healthz` probe).

File: pkg/bootkube/status.go (hunk 3 above)
Comment (mfojtik) [reply pkg/bootkube/status.go]: ack.

**Comment by mfojtik** [comment]
LGTM.

**Comment by aaronlevy** [comment]
coreosbot run e2e

**Comment by aaronlevy** [comment]
coreosbot run e2e calico

**Comment by rphillips** [comment]
coreosbot run e2e flannel

**Comment by rphillips** [comment]
coreosbot run e2e

**Comment by rphillips** [comment]
coreosbot run e2e

**Comment by sttts** [comment]
All green. Was this a flake?

**Comment by rphillips** [comment]
@sttts correct... All green and the error was a flake.

**Pull request merged**